### ISSUE TITLE
[FEATURE REQUEST] Archive metadata

### ISSUE THREAD

**Issue by btjanaka** [icaros-usc/pyribs#87]
### Description

Sometimes, we would like to store information about solutions in the archive, and this data is not necessarily a BC or objective. For instance, in the Lunar Lander example, it would be nice if we could store the total number of timesteps of the run, or the final x position. To do this, we can modify the archive API to support storing metadata. For simplicity, metadata will take the form of arbitrary objects (often a `dict`). Metadata will always be present in the archive, but it will default to None, such that users do not need to provide it. The following API changes are proposed:

* ~`__init__` gains a `use_metadata` parameter where the user can specify whether or not to use metadata. This defaults to False for backward compatibility and simplicity.~
* `initialize` allocates an object array for storing the metadata
* `add` gains a parameter for passing in a metadata object. This defaults to None ~and is ignored unless metadata is turned on.~
* `elite_with_behavior` and `get_random_elite` additionally return metadata ~(only if applicable)~
* `as_pandas` gets a parameter for including metadata (defaults to False) and adds a metadata column onto the dataframe. The resulting dataframe cannot be saved with `to_csv`, but that is something we can leave to users. Users may also just save with `to_pickle`

The `ask` and `tell` methods in the emitters and `Optimizer` must also be modified to support the metadata by taking in an array/list of objects.

Usage example:

```python
archive = GridArchive([10,10], [(-1,1)] * 2)
archive.initialize(10)  # In addition to the regular arrays, allocates a metadata array
archive.add(
    solutions=np.ones(10),
    objective_value=1.0,
    behavior_values=np.ones(2),
    metadata={  # An arbitrary metadata object (in this case a dict).
        "metadata1": np.ones(12), 
        "metadata2": 1,
    },
)
```

**Comment by btjanaka** [comment]
For simplicity, let's just make the metadata be an object. This will make the initialization much simpler, as users won't have to figure out how to convert all their data to arrays and can just pass in a dict or other object. This also makes things simpler for us. The downside is that `as_pandas` will not save the metadata nicely to a CSV file, but that does not seem very important -- users can figure out some serialization on their own or save a pickle file.

**Comment by btjanaka** [comment]
Resolved by #103, #114, and #115
